src/lib.rs

```diff
@@ -118,6 +118,47 @@ pub struct NonEmpty<T> {
     pub tail: Vec<T>,
 }
 
+pub struct NonEmptyIter<'a, T> {
+    head: Option<&'a T>,
+    tail: &'a [T],
+}
+
+impl<'a, T> Iterator for NonEmptyIter<'a, T> {
+    type Item = &'a T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some(value) = self.head.take() {
+            Some(value)
+        } else if let Some((first, rest)) = self.tail.split_first() {
+            self.tail = rest;
+            Some(first)
+        } else {
+            None
+        }
+    }
+}
+
+impl<'a, T> DoubleEndedIterator for NonEmptyIter<'a, T> {
+    fn next_back(&mut self) -> Option<Self::Item> {
+        if let Some((last, rest)) = self.tail.split_last() {
+            self.tail = rest;
+            Some(last)
+        } else if let Some(first_value) = self.head.take() {
+            Some(first_value)
+        } else {
+            None
+        }
+    }
+}
+
+impl<'a, T> ExactSizeIterator for NonEmptyIter<'a, T> {
+    fn len(&self) -> usize {
+        self.tail.len() + if self.head.is_some() { 1 } else { 0}
```

Comment:
```suggestion
        self.tail.len() + self.head.map_or(0, |_| 1)
```
Just a small suggestion :)